### PR TITLE
Epic4 Additions

### PR DESCRIPTION
**Pull request by Averynder**
This pull request touches on the audits table which affects user stories:
5.1([#62](https://github.com/Fruity-Loops/ALTA/issues/62)), 5.2([#63](https://github.com/Fruity-Loops/ALTA/issues/63))
as well as completes user stories:
3.10([#194](https://github.com/fruity-loops/alta/issues/194)) and 4.9([#132](https://github.com/Fruity-Loops/ALTA/issues/132))

**Action endpoints**

- Added API endpoint for audits:
- /proper_audits : will now return the listing as shown in the audits table
- /audit/pk/: Audits can now be deleted
- login-mobile-pin/: Now used to login once with the PIN once it has been changed

**Front-end**

- Attached the functionality for discarding / deleting audits
- Completed Audits are now changed to Active and visible in the audits table
- The audits table now displays the active audits
- Reviewing audits now has the reviewer see the listings per location

**Mobile**

- The stock keepers must now enter their email and use the pin they receive by it to log in

**Refactoring**

- Logging in has now been moved to its own app

**Miscellaneous** 

- The Script used to execute e2e tests, linting and tests has been re written to now be able to accept multiple arguments rather then one at a time as well as it now has django linters fixed for devs on windows
- Logging options have been added which now allow alot of customization as the dev now has choices on what to receive on their console input with the highest level being able to see requests and responses per API call, imported libraries (JOBS) now have the option to be disabled as well all done from the environment file

**Environment file updates**

- The following have been added to the environment file

- DEFAULT_LOG_LEVEL= This sets the level of log messages you want to receive in your terminal for any default log statements located within the application, currently used only by the jobs printing if they worked or not at a DEBUG level
- DEFAULT_LOG_HANDLER= This option decides how you want to handle the logs, the options are located within the example file however they can be set to OFF by selecting the handler or ORIGINAL to go with the original logging option
- JOB_LOG_LEVEL= similar to the default level this controls the logged messages from the app scheduler usually printing the run time of the job
- JOB_LOG_HANDLER= similar to the default log handler except for the app scheduler
- DJANGO_LOG_LEVEL= This controls the level of the default DJANGO logged output
- DJANGO_HANDLER= set to DJANGO_DEFAULT for the original terminal output
- CUSTOM_LOGGING= Set to True will enable the custom logging set up for viewsets to display request and response per API call eliminating many possible print statements for debugging
- CUSTOM_LOGGING_LEVEL= the level is customizable and if only 400 or 500 errors are wanted to be logged they can be set by WARNING for 400+ and ERROR for 500

### CODE REVIEW

File: server/audit/views.py [hunk 1]
```diff
@@ -64,26 +82,49 @@ def progression_metrics(self, request): # pylint: disable=no-self-use
...
+    for item in items:
+        popped_bin = item.pop('Bin')
+        if not dictionary['bin']:
+            dictionary['bin'] = popped_bin
+        else:
+            if popped_bin != dictionary['bin']:
+                dictionary['bin'] = 'Multiple'
+                break
+    return dictionary
+
```
Comment (NicholasNagy) [inline server/audit/views.py:103]: I don't get what the use of all this is

File: server/login/views.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    msg.attach(part1)
+    msg.attach(part2)
+
+    # Send the message via local SMTP server.
+    try:
+        mailserver = smtplib.SMTP('smtp.gmail.com', 587)
+        # identify ourselves to smtp gmail client
+        mailserver.ehlo()
+        # secure our email with tls encryption
+        # mailserver.connect('smtp.gmail.com', 465)
```
Comment (NicholasNagy) [inline server/login/views.py:162]: Please delete this commented out line of code, if we aren't going to use it

**Review by NicholasNagy — COMMENTED** [review]
I left some review comments. 

I have mostly questions, but I do have a few notes that I want to add:

1. Creating an audit is broken if there are multiple items with the same batch number
2. Despite having audits already initiated previously, I don't see anything when go to the audits tab in the sidenav
3. It's logging everything twice, which I find quite confusing, not sure if we can do anything about that.
4. What is a proper audit?

**Comment by Amanda-Kolopa** [comment]
Side note: I locally ran the backend tests and everything is passing. We will look into why they're failing for the pull_request and push builds

File: server/login/views.py (hunk 2 above)
Comment (Averynder) [reply server/login/views.py]: Okay, good call, it's been deleted!

File: server/audit/views.py (hunk 1 above)
Comment (Averynder) [reply server/audit/views.py:103]: This formats the data properly for the client side to display the information as requested in the mockups rather then sending an excess of unnecessary data and having the client filter through them

File: server/audit/views.py [hunk 3]
```diff
@@ -47,8 +49,22 @@ def list(self, request):
...
+    @action(detail=False, methods=['GET'], name='Return Proper Audit')
+    def proper_audits(self, request):
+        params = list(request.GET.keys())
+        for bad_key in ['page', 'page_size']:
+            if bad_key in params:
+                params.remove(bad_key)
+        for key in params:
+            self.queryset = self.queryset.filter(**{key: request.GET.get(key)})
+        serializer = get_proper_serializer(self.queryset, many=True)
+        data = copy.deepcopy(serializer.data)
```
[the remark this answers is not in the document]
Comment (Averynder) [reply server/audit/views.py]: good catch, removed

File: server/audit/views.py [hunk 4]
```diff
@@ -103,6 +142,12 @@ def get_item_serializer(*args, **kwargs): # pylint: disable=no-self-use
     serializer_class = ItemSerializer
     return serializer_class(*args, **kwargs)
 
+
+def get_proper_serializer(*args, **kwargs):
+    serializer_class = ProperAuditSerializer
+    return serializer_class(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (Averynder) [reply server/audit/views.py]: good catch, implemented and for the line below as well!

File: mobile/ios/App/App/capacitor.config.json [hunk 5]
```diff
@@ -1,17 +1 @@
...
-    "SplashScreen": {
-      "launchShowDuration": 0
-    }
-  },
-  "cordova": {},
-  "server": {
-    "url": "http://192.168.2.18:8100",
-    "cleartext": true
-  }
-}
```
[the remark this answers is not in the document]
Comment (Averynder) [reply mobile/ios/App/App/capacitor.config.json]: good catch, added back!

**Comment by Averynder** [comment]
> I left some review comments.
> 
> I have mostly questions, but I do have a few notes that I want to add:
> 
> 1. Creating an audit is broken if there are multiple items with the same batch number
> 2. Despite having audits already initiated previously, I don't see anything when go to the audits tab in the sidenav
> 3. It's logging everything twice, which I find quite confusing, not sure if we can do anything about that.
> 4. What is a proper audit?

1. I'm not sure how you are getting this but the batch number is supposed to be a unique id used for only one item. You can verify this in the model, nothing in this PR has touched the inventory items
2. Only Active Audits are shown belonging to your organization, it's possible you have incorrect test data, you can verify this working when using the fixtures. I can create an audit for both an IM and SA and it works completely fine, you would have to provide the exact data as I am unable to reproduce the test case
3. Logging was tested on both Windows and Linux machines, it could be possible there is a bug with Mac however I believe it may just be your environment file is improperly configured, I have left notes now in the PR description
4. A proper audit is just the name given to have the audit formatted as the mockups show where the information is already transformed

**Comment by Averynder** [comment]
> Side note: I locally ran the backend tests and everything is passing. We will look into why they're failing for the pull_request and push builds

I had a similar issue, ran the complaining tests manually and via script as well as our execute script and couldn't reproduce a failure at all, even when trying the api tests individually

File: server/audit/views.py (hunk 1 above)
Comment (NicholasNagy) [reply server/audit/views.py:103]: awesome, okay, I see

**Comment by NicholasNagy** [comment]
> > I left some review comments.
> > I have mostly questions, but I do have a few notes that I want to add:
> > 
> > 1. Creating an audit is broken if there are multiple items with the same batch number
> > 2. Despite having audits already initiated previously, I don't see anything when go to the audits tab in the sidenav
> > 3. It's logging everything twice, which I find quite confusing, not sure if we can do anything about that.
> > 4. What is a proper audit?
> 
> 1. I'm not sure how you are getting this but the batch number is supposed to be a unique id used for only one item. You can verify this in the model, nothing in this PR has touched the inventory items
> 2. Only Active Audits are shown belonging to your organization, it's possible you have incorrect test data, you can verify this working when using the fixtures. I can create an audit for both an IM and SA and it works completely fine, you would have to provide the exact data as I am unable to reproduce the test case
> 3. Logging was tested on both Windows and Linux machines, it could be possible there is a bug with Mac however I believe it may just be your environment file is improperly configured, I have left notes now in the PR description
> 4. A proper audit is just the name given to have the audit formatted as the mockups show where the information is already transformed

Makes sens to me, thank you for the responses!

**Comment by NicholasNagy** [comment]
I also fixed the broken dependencies in our tests and the code duplicates.

**Review by Amanda-Kolopa — APPROVED** [review]
Thank you guys for all the modifications!
This PR looks good to merge :smile:
